relay-server/src/actors/project_cache.rs

```diff
@@ -0,0 +1,250 @@
+use crate::actors::upstream::UpstreamRelay;
+use crate::actors::upstream_project_source::UpstreamProjectSource;
+use crate::metrics::{RelayCounters, RelayHistograms, RelayTimers};
+use crate::utils::{RedisPool, Response};
+
+#[cfg(feature = "processing")]
+use {crate::actors::redis_project_source::RedisProjectSource, relay_common::clone};
+
+#[derive(Fail, Debug)]
+pub enum ProjectError {
+    #[fail(display = "failed to fetch project state from upstream")]
+    FetchFailed,
+
+    #[fail(display = "could not schedule project fetching")]
+    ScheduleFailed(#[cause] MailboxError),
+}
+
+impl ResponseError for ProjectError {}
+
+struct ProjectEntry {
+    last_updated_at: Instant,
+    project: Addr<Project>,
+}
+
+pub struct ProjectCache {
+    config: Arc<Config>,
+    projects: HashMap<ProjectId, ProjectEntry>,
+
+    local_source: Addr<LocalProjectSource>,
+    upstream_source: Addr<UpstreamProjectSource>,
+    #[cfg(feature = "processing")]
+    redis_source: Option<Addr<RedisProjectSource>>,
+}
+
+impl ProjectCache {
+    pub fn new(
+        config: Arc<Config>,
+        upstream_relay: Addr<UpstreamRelay>,
+        _redis: Option<RedisPool>,
+    ) -> Self {
+        let local_source = LocalProjectSource::new(config.clone()).start();
+        let upstream_source = UpstreamProjectSource::new(config.clone(), upstream_relay).start();
+
+        #[cfg(feature = "processing")]
+        let redis_source = _redis.map(|pool| {
+            SyncArbiter::start(
+                config.cpu_concurrency(),
+                clone!(config, || RedisProjectSource::new(
+                    config.clone(),
+                    pool.clone()
+                )),
+            )
+        });
+
+        ProjectCache {
+            config,
+            projects: HashMap::new(),
+
+            local_source,
+            upstream_source,
+            #[cfg(feature = "processing")]
+            redis_source,
+        }
+    }
+
+    fn evict_stale_project_caches(&mut self) {
+        metric!(counter(RelayCounters::EvictingStaleProjectCaches) += 1);
+        let eviction_start = Instant::now();
+        let eviction_threshold = eviction_start - 2 * self.config.project_cache_expiry();
+        self.projects
+            .retain(|_, entry| entry.last_updated_at > eviction_threshold);
+
+        metric!(timer(RelayTimers::ProjectStateEvictionDuration) = eviction_start.elapsed());
+    }
+}
+
+impl Actor for ProjectCache {
+    type Context = Context<Self>;
+
+    fn started(&mut self, context: &mut Self::Context) {
+        // Set the mailbox size to the size of the event buffer. This is a rough estimate but
+        // should ensure that we're not dropping messages if the main arbiter running this actor
+        // gets hammered a bit.
+        let mailbox_size = self.config.event_buffer_size() as usize;
+        context.set_mailbox_capacity(mailbox_size);
+
+        context.run_interval(2 * self.config.project_cache_expiry(), |slf, _| {
```

Comment:
We might want to run this more often to reduce the impact of a single run. We could even hard-code it to 60s.

Reply:
does `project_cache_expiry / 2` make sense?

Reply:
Hm, I think no more than this. Conceptually, the expiry poll is independent of the interval and more similar to the local state refresh interval.

Reply:
I added a configuration param that defaults to 1 minute